_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,192 @@
+
+Let's next have a look at the example of DSL of one of the sample rules presented in the previous post:
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rule-dsl.png" width="800px" alt="Figure 1: Rules DSL"/>
+<br/>
+<i><small>Figure 1: Rules DSL</small></i>
+</center>
+<br/>
+![](assets/rule-dsl.png)
+
+Earlier we have already seen how part of this definition, namely `groupingKeyNames`, is used in `DynamicKeyFunction` to extract message keys. At the same time, all other parameters of this rule are necessary within `DynamicAlertFunction` - they define the actual logic of performed operations and it's parameters (such as the alert triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use a special data distribution mechanism in Apache Flink - [events broadcast](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/state/broadcast_state.html). Let’s turn to the Job Graph of the system that we are building:
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph"/>
+<br/>
+<i><small>Figure 2: Job Graph</small></i>
+</center>
+<br/>
+![](assets/job-graph.png)
+
+On it you can see the main blocks of our Transactions processing pipeline:<br>
+
+* **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>
+    * The forward operator following it means that all data consumed by one of the parallel instances of Transaction Source operator will be transferred to exactly one instance of the subsequent DynamicKeyFunction operator, without redistribution. It also indicates the same level of parallelism of two connected operators (12 in this case).
+
+ <center>
+ <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/forward.png" width="800px" alt="Figure 3: Message passing across operators : FORWARD"/>
+ <br/>
+ <i><small>Figure 3: Message passing across operators : FORWARD</small></i>
+ </center>
+ <br/>
+![](assets/forward.png)
+
+* **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator performs `hashing` of this dynamic key and corresponding data sharding
+    * hash here means that for each message a hash code is calculated and messages are evenly distributed among available parallel instances of the following operator.
+
+ <center>
+ <img src="{{ site.baseurl }}/img/blog/patterns-blog-2/hash.png" width="800px" alt="Figure 3: Message passing across operators : HASH (keyBy)"/>
```

Comment:
We definitely need to sync on the task/operator terminology ;)
I use the same visual encoding as in the first part of the blog - circles are transactions (https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html)
On this image dotted lines are tasks ("Tasks encapsulate exactly one parallel instance of an Operator or Operator Chain."). Maybe I should also add links to the Glossary when I first mention "task" and "operator".

Reply:
Sure, you could add links if you want to. In the end, when reading this post, I think, readers would, however take whatever term you present as fore-granted. Nevertheless it should be the correct one ;)
Think about the simpler image but you could also keep (and adapt) the current one which conveys the message but takes some more space. If you keep the current one, you may also need to think about adding the remaining arrows because otherwise the image is incomplete.

Reply:
@NicoK which arrows do you mean? Currently each event (orange circle) is dispatched according to it's hash code - I do not see which additional errors could be necessary.

Reply:
I actually mixed up what the circles meant. If it is a transaction/event, it is ok to keep the arrows as they are (I thought they indicate all communication paths and are missing some).
In that case, Operator_1_2's event may miss its hash. Does it make sense to draw dotted lines for the paths which are not shown? Or is that cluttering the image too much?

The proposal you did not talk about though is whether it makes sense to simplify the image by having Operator_1 with p=1 only and instead two events in there.

Reply:
I have added a sentence with explicit explanation, so that people also do not get confused.
So, regarding the hash - I actually gave it a try when I was drawing it. It did not fit on the right because of all the arrows and I decide not to bother. If you think it unclear like that, I can move the hash "bubble" to the left. 
Regarding p=1 
- I though about about it, but, as you said, I need two circles within the same task - this makes it stand out from all the other figures, I did not like that idea. Additionally, I feel that this variant looses part of the message that this process happens in parallel in every task.
- Then I though about having p = 2, this also makes things awkward: we either forward both circles to the same receiving end, which leaves one just hanging there, or forward them to one respectively, either making them look like a forward, or a bow tie, which somehow also makes it easy to miss the point.
This is how I came to this variant. I think it is the best option. I am open to adding the missing "hash" if you think it is important.

Reply:
no, the "hash" is not too important; also thanks for the explanation regarding p=1 - makes sense